locator/topology.cc

```diff
@@ -595,8 +599,15 @@ void topology::do_sort_by_proximity(T address, utils::small_vector<T, 3>& addres
         return i1.distance < i2.distance;
     });
     auto it = addresses.begin();
-    for (const auto& i : host_infos) {
-        *it++ = i.id;
+    for (size_t i = 1; i < host_infos.size(); ++i) {
+        // Shuffle equal-distance replicas to improve load-balancing
+        if (host_infos[i].distance == host_infos[i-1].distance) {
+            if (shuffler & 1) {
+                std::swap(host_infos[i], host_infos[i-1]);
+            }
+            shuffler = std::rotr(shuffler, 1);
+        }
+        *it++ = host_infos[i-1].id;
```

Comment:
This is executed n-1 times, in the previous code n times.

Reply:
You're right, the last component must be copied too